src/HTTPProxy.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+class HTTPProxy implements \danog\MadelineProto\Proxy {
+
+    private $sock;
+    private $protocol;
+    private $timeout = ['sec' => 0, 'usec' => 0];
+    private $domain;
+    private $type;
+    private $options = [];
+    private $use_connect = FALSE;
+    private $use_ssl = FALSE;
+
+    public function __construct($domain, $type, $protocol) {
+        $this->domain = $domain;
+        $this->type = $type;
+        $this->protocol = $protocol === 'https' ? 'tls' : 'tcp';
+
+        if ($protocol !== 'http') {
+            $this->use_connect = TRUE;
+            if ($protocol === 'https') {
+                $this->use_ssl = TRUE;
+            }
+        }
+    }
+
+    public function __destruct() {
+        if ($this->sock !== null) {
+            fclose($this->sock);
+            $this->sock = NULL;
+        }
+    }
+
+    public function accept() {
+        throw new \danog\MadelineProto\Exception('Not supported');
+    }
+
+    public function bind($address, $port = 0) {
+        throw new \danog\MadelineProto\Exception('Not supported');
+    }
+
+    public function close() {
+        fclose($this->sock);
+        $this->sock = null;
+    }
+
+    public function connect($address, $port = 0) {
+        $errno = 0;
+        $errstr = '';
+
+        if (isset($this->options['host']) && isset($this->options['port'])) {
+            $this->sock = @fsockopen($this->options['host'], $this->options['port'], $errno, $errstr, $this->timeout['sec'] + ($this->timeout['usec'] / 1000000));
```

Comment:
You really should use the \Socket class instead

Reply:
\Socket class cannot use https as transport protocol, so there's still need for https proxy class like \FSocket.

to use sockets when you make direct connections to Telegram servers may gain some performance in some scenarios, but this gain it's lost when you have a proxy in between.

On the other hand, if \Socket class where to use PHP Streams instead of PHP Sockets, probably there wouldn't be need for a separate Proxy Class.